src-exe/cabal-plan.hs

```diff
@@ -182,6 +190,42 @@ doDot showBuiltin showGlobal plan = do
 
     putStrLn "}"
 
+doTopo :: Bool -> Bool -> PlanJson -> IO ()
+doTopo showBuiltin showGlobal plan = do
+    let units0 = pjUnits plan
+        units1 | showBuiltin = units0
+               | otherwise   = M.filter ((UnitTypeBuiltin /=) . uType) units0
+        units2 | showGlobal  = units1
+               | otherwise   = M.filter ((UnitTypeGlobal  /=) . uType) units1
+        units = units2
+
+    let g :: Alga.Graph Unit
+        g = Alga.edges
+            [ (unit, dep)
+            | unit <- toList units
+            , depId <- toList $ foldMap (\ci -> ciLibDeps ci <> ciExeDeps ci) (uComps unit)
+            , depId /= uId unit
+            , dep <- maybeToList $ M.lookup depId units
+            ]
+
+        am :: AM.AdjacencyMap UnitId
+        am = AlgaC.toGraph (fmap uId g)
+
+    for_ (AM.topSort am) $ \xs -> for_ xs $ \unitId ->
```

Comment:
It looks like you wouldn't need to use `AlgaC` or `AM` if `Algebra.Graph` exported `topSort` on the `Graph` datatype. This is easy to implement and would make the code here much nicer.

Reply:
can there be `topSort` on `Graph a`, without `Ord a`, that would be great.

Reply:
Hmm, interesting. `Eq a` is definitely required. Without `Ord a` I'm afraid `topSort` will become quite expensive, because we essentially will have to implement the `Eq a`-based `nub` functionality: given a graph expression `a + d + a + c + b + a + b` we'll need to produce the list `[a, d, c, b]`.

In short: yes, we can, but it will be too slow to be useful.